### PR TITLE
feat: integrate openID auth logic and user profile management

### PR DESCRIPTION
**Pull request by alexstan12**
- this change introduces OpenID authn by using providers such as Github, Gitlab and Google. User sessions are now used for web clients to identify and persist an authenticated users session, thus not requiring every request to use credentials.
- UserDB added that contains users security data.
- dex service introduced to replace multiple providers implementations

Signed-off-by: Alex Stan <alexandrustan96@yahoo.ro>

**What type of PR is this?**

**Which issue does this PR fix**:

**What does this PR do / Why do we need it**:

**If an issue # is not available please add repro steps and logs showing the issue**:

**Testing done on this change**:

**Automation added to e2e**:

**Will this break upgrades or downgrades?**

**Does this PR introduce any user-facing change?**:

```release-note

```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/api/routes.go [hunk 1]
```diff
@@ -134,6 +162,104 @@ func (rh *RouteHandler) SetupRoutes() {
 	}
 }
 
+func (rh *RouteHandler) UserInfoLogic(ctlr *Controller) rp.CodeExchangeUserinfoCallback {
```
Comment (andaaron) [inline pkg/api/routes.go:165]: Should be after comment `// Method handlers`, and should be documented, just like the other RouteHandlers below. But let's discuss these endpoints with @rchincha and @peusebiu first.

File: pkg/api/routes.go [hunk 2]
```diff
@@ -149,9 +275,16 @@ func (rh *RouteHandler) CheckVersionSupport(response http.ResponseWriter, reques
...
 	if rh.c.Config.HTTP.Auth != nil {
 		if rh.c.Config.HTTP.Auth.Bearer != nil {
-			response.Header().Set("WWW-Authenticate", fmt.Sprintf("bearer realm=%s", rh.c.Config.HTTP.Auth.Bearer.Realm))
+			response.Header().Add("WWW-Authenticate", fmt.Sprintf("bearer realm=%s", rh.c.Config.HTTP.Auth.Bearer.Realm))
 		} else {
-			response.Header().Set("WWW-Authenticate", fmt.Sprintf("basic realm=%s", rh.c.Config.HTTP.Realm))
+			response.Header().Add("WWW-Authenticate", fmt.Sprintf("basic realm=%s", rh.c.Config.HTTP.Realm))
+		}
+
+		authenticateHeaders := removeHeadersDuplicates(response.Header().Values("WWW-Authenticate"))
```
Comment (andaaron) [inline pkg/api/routes.go:283]: Why were the changes in this function needed?

File: examples/README.md [hunk 3]
```diff
@@ -172,6 +179,16 @@ Should authentication fail, to prevent automated attacks, a delayed response can
...
+#### API keys
+
+Use it to activate API keys
+
+```
+  "http": {
+    "auth": {
+      "apikeys": true
+```
+
```
Comment (andaaron) [inline examples/README.md:191]: Please also add config file under examples with zot configuration for dex.

Also, we need steps, not sure in what document, on how to configure the dex service itself to work with zot.

File: README_APIkeys.md [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+### Create API Key
+**Description**: Create an API key for the current user. Returns an error if API key already exists - use regenerate API key instead.
+
+**Usage**: POST /api/security/apiKey
+
+**Produces**: application/json
+
+**Sample input**:
+```
+POST /api/security/apiKey
```
[the remark this answers is not in the document]
Comment (rchincha) [reply README_APIkeys.md:30]: Yes, let's setup a meeting to talk about this

File: README_APIkeys.md
[the remark this answers is not in the document]
Comment (rchincha) [reply README_APIkeys.md:1]: under the feature is fine

File: README_APIkeys.md [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+**Usage**: DELETE /api/security/apiKey
+
+**Produces**: application/json
+
+### Revoke User API Key
+**Description**: Revokes the API key of another user
+
+**Security**: Requires a privileged user (Admin only)
+
+**Usage**: DELETE /api/security/apiKey/{username}
```
Comment (andaaron) [inline README_APIkeys.md:72]: /v2/_zot/mgmt/security/apiKey

**Comment by andaaron** [comment]
Superseded by https://github.com/project-zot/zot/pull/1381

**Pull request closed without merging**